lib/component.js

```diff
@@ -805,6 +819,114 @@ class Component extends WebComponent {
     return this._attrs;
   }
 
+  _syncParams() {
+    // maintain local validated map where all schema keys are defined
+    this._paramTypes = {};
+    const paramTypes = this.getConfig(`params`);
+
+    // vuejs like this.data.$data to reference "real" value of params
+    Object.defineProperty(this.params, `$params`, {
+      enumerable: false,
+      value: {},
+    });
+
+    for (const param of Object.keys(paramTypes)) {
+      // convert type shorthand to object
+      let paramType = paramTypes[param];
+      if (PARAM_TYPE_DEFAULTS.has(paramType)) {
+        paramType = {type: paramType};
+      }
+
+      // Ensure param type is valid
+      const type = paramType.type;
+      if (!PARAM_TYPE_DEFAULTS.has(type)) {
+        const typeString = typeof type === `function` ? type.name : String(type);
+        throw new Error(
+          `Invalid type: ${typeString} for attr: ${param} in paramType. ` +
+            `Only (${Array.from(PARAM_TYPE_DEFAULTS.keys())
+              .map((v) => `'${v.name}'`)
+              .join(` | `)}) is valid.`,
+        );
+      }
+
+      if (paramType.default && paramType.required) {
+        throw new Error(`${this}: param '${param}' cannot have both required and default`);
+      }
+
+      if ([Object, Array, Map, Set].includes(type) && !paramType.shouldUpdate) {
+        throw new Error(
+          `Object like params '${param}' on ${this.constructor.name} must have 'shouldUpdate' function passed in`,
+        );
+      }
+
+      const paramSchemaObj = {
+        type,
+        default: paramType.default ? paramType.default : PARAM_TYPE_DEFAULTS.get(type),
+        required: paramType.required ? paramType.required : false,
+        // by default use === for comparision
+        shouldUpdate: paramType.shouldUpdate || ((oldVal, newVal) => oldVal !== newVal),
```

Comment:
There's no need to declare & bind a new arrow function `(oldVal, newVal) => oldVal !== newVal` for every param that doesn't define `shouldUpdate`. There should be one non-arrow function `function defaultShouldUpdate(oldVal, newVal) { return oldVal !== newVal; }` declared once which is then reused each time it's needed.